src/background_iterator.rs

```diff
@@ -0,0 +1,149 @@
+use std;
+use std::sync::mpsc::{sync_channel, Receiver};
+use std::thread::{spawn, JoinHandle};
+
+/// Execute an iterator on a worker thread, which can work ahead a configurable number of items.
+pub struct BackgroundIterator<T> {
+    rx: Receiver<Option<T>>,
+    done: bool,
+    handle: Option<JoinHandle<()>>,
+}
+
+impl<T: Send> Iterator for BackgroundIterator<T> {
+    type Item = T;
+
+    fn next(&mut self) -> Option<T> {
+        if self.done {
+            return None;
+        }
+
+        match self.rx.recv() {
+            Ok(Some(v)) => Some(v),
+            Ok(None) => {
+                self.done = true;
+                None
+            }
+            // if the producer thread dies, then the sender thread must have panicked
+            // propagate the panic, otherwise we will silently continue with likely incomplete results
+            Err(_) => {
+                self.done = true;
+                let r = self.handle.take().unwrap().join();
+                match r {
+                    Ok(()) => {
+                        // this state should be unreachable,
+                        // but presumably benign if we get here
+                        None
+                    }
+                    Err(e) => {
+                        // sender panicked - we will re-panic the panic.
+                        // need to do a little dance here to get the panic message
+                        let msg = match e.downcast_ref::<&'static str>() {
+                            Some(s) => s.to_string(),
+                            None => match e.downcast_ref::<String>() {
+                                Some(s) => s.clone(),
+                                None => {
+                                    "unknown panic on BackgroundIterator worker thread".to_string()
+                                }
+                            },
+                        };
+                        panic!(msg);
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl<T: 'static + Send> BackgroundIterator<T> {
+    /// Iterate through `itr` on a newly created thread, and send items back to the returned
+    /// `BackgroundIterator` for consumption on the calling thread. The worker thread will
+    /// continue to produce elements until it is `max_read_ahead` items ahead of the consumer iterator.
+    pub fn new<I: 'static + Send + Iterator<Item = T>>(
+        itr: I,
+        max_read_ahead: usize,
+    ) -> BackgroundIterator<T> {
+        let (tx, rx) = sync_channel::<Option<T>>(max_read_ahead);
+        let handle = spawn(move || {
+            for item in itr {
+                if tx.send(Some(item)).is_err() {
+                    return;
+                };
+            }
+
+            tx.send(None).unwrap();
+        });
+
+        BackgroundIterator {
+            rx,
+            handle: Some(handle),
```

Comment:
What I was saying with my earlier comment is I'm not quite sure under what circumstances it will try to join on this handle.  Will it try to join when the `BackgroundIterator` is destroyed?  The situation I'm worried about is:
1. The reader thread panics
2. While unwinding the stack, it tries to destroy `BackgroundIterator`, so it waits on the thread, but
3. The thread is blocking on `tx.send`, which won't return an error until `rx` is cleaned up, which it won't be until after the destructor call completes, which it won't because it's waiting on that thread.

Reply:
From JoinHandle docs: "A JoinHandle detaches the associated thread when it is dropped, which means that there is no longer any handle to thread and no way to join on it."

The `rx` will be dropped on the reader thread, which will cause the next `tx.send` to return an Error, and the sender thread will shut down.